[#335] Navigation to Operation#securityScheme, Link#operationId, and …

ghillairet opened the pull request:
…Link#operationRef

See #335 

Added new hyperlink detectors for security schemes and link operationId. For link operationRef, I modified the existing JsonReferenceDetector to work not only on $ref but also on operationRef. 
Added tests for security schemes and link operationId detectors.

ghillairet committed:
[#335] Navigation to Operation#securityScheme, Link#operationId, and Link#operationRef

tfesenko commented on com.reprezen.swagedit.core/src/com/reprezen/swagedit/core/hyperlinks/JsonReferenceHyperlinkDetector.java:
```diff
@@ -37,7 +37,7 @@
     @Override
     protected boolean canDetect(JsonPointer pointer) {
-        return pointer != null && pointer.toString().endsWith("$ref");
+        return pointer != null && (pointer.toString().endsWith("$ref") || pointer.toString().endsWith("operationRef"));
```
`(pointer.toString().endsWith("$ref")` is a standard reference, while the one which `.endsWith("operationRef")` is not. How about extracting the check for the "operationRef" in a hook method to make it clear that it's a customization?

tfesenko commented on com.reprezen.swagedit.core/src/com/reprezen/swagedit/core/json/references/JsonReferenceFactory.java:
```diff
@@ -114,6 +114,11 @@ public JsonReference doCreate(String value, Object source) {
         String fragment = uri.getFragment();
         JsonPointer pointer = null;
         try {
+            // Pointer fails to resolve if ends with /
```
Thanks for the explanatory comment!

tfesenko commented on com.reprezen.swagedit.openapi3.tests/src/com/reprezen/swagedit/openapi3/editor/hyperlinks/HyperlinkDetectorTest.xtend:
```diff
@@ -0,0 +1,83 @@
+package com.reprezen.swagedit.openapi3.editor.hyperlinks
```
A copyright header?

tfesenko replied:
The same for `HyperlinkDetectorTest`, `LinkOperationHyperlinkDetector `, and  `SecuritySchemeHyperlinkDetector `

tfesenko commented on com.reprezen.swagedit.openapi3.tests/src/com/reprezen/swagedit/openapi3/editor/hyperlinks/HyperlinkDetectorTest.xtend:
```diff
@@ -0,0 +1,83 @@
+	val operationDetector = new LinkOperationHyperlinkDetector
+
+	@Test
+	def void testShouldCreateHyperLink_FromSecuirtyRequirement() throws Exception {
```
a typo in `Secu**ir**ty`

tfesenko commented on com.reprezen.swagedit.openapi3.tests/src/com/reprezen/swagedit/openapi3/editor/hyperlinks/HyperlinkDetectorTest.xtend:
```diff
@@ -0,0 +1,83 @@
+			document.getLineOfOffset(groups.get("2").offset),
+			document.getLineOfOffset(target.offset)
+		)
+	}
```
Thanks!

tfesenko commented on com.reprezen.swagedit.openapi3.tests/src/com/reprezen/swagedit/openapi3/utils/Cursors.xtend:
```diff
@@ -12,9 +12,18 @@ import org.eclipse.xtext.xbase.lib.Functions.Function2
 import org.eclipse.xtext.xbase.lib.Procedures.Procedure2
```
A missing copyright header here

tfesenko commented on com.reprezen.swagedit.openapi3/src/com/reprezen/swagedit/openapi3/hyperlinks/LinkOperationHyperlinkDetector.java:
```diff
@@ -0,0 +1,52 @@
+    @Override
+    protected boolean canDetect(JsonPointer pointer) {
+        return pointer != null && pointer.toString().matches("/components/links/(\\w+)/operationId");
+    }
```
This regex only works when a link is defined in `components`, but it can also be defined directly in a response:
```yaml
  /2.0/repositories/{username}/{slug}: 
    get: 
      operationId: getRepository
      responses: 
        '200':
          description: The repository
          links:
            repositoryPullRequests:
              operationId: getRepository
```
In that case, navigation does not work.

ghillairet committed:
Merge branch 'master' into task/335

# Conflicts:
#	com.reprezen.swagedit.openapi3.tests/src/com/reprezen/swagedit/openapi3/utils/Mocks.java

ghillairet committed:
[#335] Navigation to Operation#securityScheme, Link#operationId, and Link#operationRef

Fix tests

ghillairet committed:
[#335] Navigation to Operation#securityScheme, Link#operationId, and Link#operationRef

Separete Detectors for JsonRef and OperationRef
Abstract class ReferenceHyperlinkDetector has been added, this detector is able to detect local and external references, but does not implement the canDetect method. This method is implemented by JsonReferenceHyperlinkDetector to detect references from $ref field, and by OperationRefHyperlinkDetector to detect references from operationRef field. The OperationRefHyperlinkDetector is only used in OpenApi3.

ghillairet committed:
Missing copyrights.

ghillairet committed:
[#335] Navigation to Operation#securityScheme, Link#operationId, and Link#operationRef

Fix detection of hyperlink for links inside responses.

ghillairet committed:
Missing copyrights.

tfesenko commented on com.reprezen.swagedit.openapi3.tests/src/com/reprezen/swagedit/openapi3/editor/hyperlinks/HyperlinkDetectorTest.xtend:
```diff
@@ -128,4 +128,70 @@ class HyperlinkDetectorTest {
+			        '200':
+			          links:
+			            foo:
+			              operationId: li<1>st
```
Thanks!

tfesenko commented on com.reprezen.swagedit.openapi3/src/com/reprezen/swagedit/openapi3/hyperlinks/LinkOperationRefHyperlinkDetector.java:
```diff
@@ -0,0 +1,32 @@
+    @Override
+    protected boolean canDetect(JsonPointer pointer) {
+        return pointer != null && pointer.toString().endsWith("operationRef");
+    }
```
:thumbsup:

tfesenko commented on com.reprezen.swagedit.openapi3.tests/src/com/reprezen/swagedit/openapi3/validation/ValidatorTest.xtend:
```diff
@@ -1,3 +1,13 @@
+ *
+ * Contributors:
+ *    ModelSolv, Inc. - initial API and implementation and/or initial documentation
+ *******************************************************************************/
```
Thanks!
